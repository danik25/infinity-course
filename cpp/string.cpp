/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}