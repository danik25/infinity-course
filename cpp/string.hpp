/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */