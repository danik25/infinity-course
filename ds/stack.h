/****************************************************************
*                       Data Structures - Stack                 *
* Related files: stack_test.c, stack.h, stack_test.c            * *															    *
* writer: Dani Benjamin                                         *
* reviewer: Ofer Tauber                                         *
* last update: 29.3.19                                          *
****************************************************************/        
       
#ifndef _STACK_H_
#define _STACK_H_

#include <stdio.h> /* size_t */

typedef struct stack stack_t;

/*creates the stack with capacitiy of <capacitiy>, for element of size
 <element_size>. returns a pointer to stack*/
stack_t *StackCreate(size_t element_size, size_t capacity);


/*destroy the stack and free the allocated memory*/
void StackDestroy(stack_t *stack_ptr);


/*insert element <element> to the top of the stack(by value).
*returns 0 to denote success or 1 to denote failure (pushing a full stack).*/
int StackPush(stack_t *stack_ptr, const void *element);


/*remove the element in top of tghe stack(without returning it's value)*/
void StackPop(stack_t *stack_ptr);


/*returns pointer to the top element without modifying the stack*/
void *StackPeek(const stack_t *stack_ptr);


/*returns the size of the stack without modifying it*/
size_t StackSize(const stack_t *stack_ptr);

#endif /*_STACK_H_*/
