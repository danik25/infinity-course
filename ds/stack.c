#include <stdio.h> /*printf*/
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/
#include <string.h> /*memcpy*/

#include "stack.h"

struct stack
{
    void *end;
    void *current;
	size_t element_size;
	void *head;
};

stack_t *StackCreate(size_t element_size, size_t capacity)
{    
    stack_t *stack = malloc(sizeof(stack_t));
    stack -> head = malloc(capacity * element_size);
    
    stack -> end = (char*)stack -> head + capacity * element_size;
    stack -> element_size = element_size;
    /*current is pointing to one element size under the head*/
    stack -> current = (char*)stack -> head - element_size;
    
    if (NULL == stack || NULL == stack->head)
    {
    	return NULL;
    }
    return stack;
}

void StackDestroy(stack_t *stack_ptr)
{
	assert(stack_ptr);
	
    free(stack_ptr -> head);
    stack_ptr -> head = NULL;
    free(stack_ptr);
    stack_ptr = NULL;
}

int StackPush(stack_t *stack_ptr, const void *element)
{
	assert(stack_ptr);
	assert(element);
	/*empty stack*/
	if (stack_ptr -> current == NULL)
	{
	    stack_ptr -> current = (char*)stack_ptr -> head - stack_ptr ->
	                                element_size;
	}
	else if (stack_ptr -> head == NULL)
	{
	    fprintf(stderr, "%s", "unsuccessfull push, no stack\n\n");
	        return 1;
	}

    else if ((char*)stack_ptr -> current + stack_ptr -> element_size == 
    								stack_ptr -> end)
    {
        fprintf(stderr, "%s", "unsuccessfull push, full stack\n\n");
     	return 1; /*full stack*/
    }
    stack_ptr -> current = (char*)stack_ptr -> current + 
    								stack_ptr -> element_size;
    memcpy(stack_ptr -> current, element, stack_ptr -> element_size);
    
    return 0;
}

void StackPop(stack_t *stack_ptr)
{
	assert(stack_ptr);
	
    if (stack_ptr -> current < stack_ptr -> head)
    {
        fprintf(stderr, "%s", "unsuccessfull pop, empty stack\n\n");
    }
    else
    {
        stack_ptr -> current = (char*)stack_ptr -> current - stack_ptr ->
                                                                element_size;
        /*if it was the last element in the stack*/
        if (stack_ptr -> current < stack_ptr -> head)
        {
        	stack_ptr -> current = NULL;
        }
    }
}

void *StackPeek(const stack_t *stack_ptr)
{
	assert (stack_ptr);
    return stack_ptr->current;
}

size_t StackSize(const stack_t *stack_ptr)
{	
	size_t size = ((char*)stack_ptr->current - (char*)stack_ptr->head +
	                        stack_ptr->element_size)/stack_ptr->element_size;
	
	assert(stack_ptr);
	
	if (stack_ptr -> current < stack_ptr -> head)
	{
		return 0;
	}
	return size;
}







