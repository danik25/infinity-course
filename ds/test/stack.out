/home/xyz/git/dani-benjamin/ds/stack.out